### ISSUE TITLE
Parsing is extremely slow for files with lots of embedded newlines

### ISSUE THREAD

**Issue by ronnieoverby** [tspence/csharp-csv-reader#21]
This csv file is 14.6 MB, has 3 columns and 113 records: [codebase.zip](https://github.com/tspence/csharp-csv-reader/files/4434997/codebase.zip)

Here's the code to read the file:
```c#
void Main()
{
	using var reader = new StreamReader(@"C:\codebase.csv");
	using var csvreader = new CSVFile.CSVReader(reader);
	Console.WriteLine(csvreader.Count());
}
```

My machine spends ~9.5 minutes parsing it using your library. Other libraries provide sub-second parsing.

**Comment by tspence** [comment]
Hmm.

If I'm reading this file correctly, codebase.csv is a CSV file with filename, size, and text.  Your column #3 includes a lot of embedded newlines.  When I was originally working on this library, I optimized for extremely large (terabyte-sized) files that didn't have embedded newlines.  The logic I used to stream text off disk relied on pulling one line at a time, verifying that the CSV was coherent, and then adding another line if it didn't end properly.

I think you've found a case where this library isn't well optimized.  I'm glad to know it still produces correct results, but this is definitely something that can be improved.

**Comment by tspence** [comment]
Looking at this bit of code, I realize now why I never fixed this performance issue.  The strains of trying to be compatible with all different versions of C# back to DotNet 2.0 means that the algorithm can't rely on some of the nice abstractions introduced in later frameworks.  

Anyways, let's go ahead and tackle this, even though I'm looking at this issue a year after you raised it, why not :)

**Comment by ronnieoverby** [comment]
I pointed this out to you because I was actively working on developing the best performing CSV parser for .net.

I noticed your library was the only one. at that time, that outperformed mine on some of my benchmarks. In other benchmarks (lots of embedded newlines) your performance was abysmal and I was just letting you know.

Personally, I hope that you don't fix it :P

**Comment by tspence** [comment]
The code is open source, so please feel free to adapt and share if you give credit :)  This was a project I wrote way back in 2007 when I was a contractor that I've gradually updated over the years; I'm just happy that people still find it useful.

Thank you for contributing!

**Comment by tspence** [comment]
Looks like the fix works.  Will be publishing version 3.0 as soon as I can get the async reader to work.

**Comment by tspence** [comment]
Believe this is now solved with the release of CSVFile 3.0.2.

Thanks for the feedback @ronnieoverby !
